### DISCUSSION TITLE
The Go Programming Rule

### DISCUSSION THREAD

**Discussion by AlexStocks** [arana-db/arana#200] [Announcements]
Go 编程规范

**Comment by AlexStocks** [comment]
* for-loop 循环过程中，value 一直是同一个变量，其值变化而已

```Go
	for table, rule := range router.mrMap {
		o.ShardingRuleMap[strings.ToLower(table)] = &rule
	}
```

截图用了 for-loop 的变量，结果导致 o.ShardingRuleMap[strings.ToLower(table)]
结果都指向了一个临时变量，效果如下：

![image](https://user-images.githubusercontent.com/7959374/170958030-d141d6fd-7e90-4f78-813a-e6e4ca5d1f14.png)

改进后代码如下：

```Go
	for table := range router.mrMap {
		rule := router.mrMap[table]
		o.ShardingRuleMap[strings.ToLower(table)] = &rule
	}
```

**Reply by ma642** [reply]
o.ShardingRuleMap[strings.ToLower(table)] = &router.mrMap[table]
不会增加新的变量和copy

**Comment by seeflood** [comment]
这个感觉可以通过linter 做检查？

**Comment by AlexStocks** [comment]
不要调用 struct.String() 函数。在记录日志过程中，不管日志是否输出，这个函数都会被调用。https://github.com/apache/dubbo-go/pull/1955
